### PR TITLE
feat: thorchain swap transaction history detection

### PR DESCRIPTION
**Pull request by kaladinlight**
- abstracts history service to support variable backend sources
- adds `HistoryTx` interface to support disparate backend source payloads (supports creating "txs" from block events)
- add concurrency to history fetching
- replace handrolled tenderrmint types with provided types utilizing tendermint json unmarshaller
- synthesizes transaction history for swap events via `outbound` event
- matches fee event with the `outbound.in_tx_id` if possible (defaults to 0.02 rune if unable to match)

### CODE REVIEW

File: go/pkg/cosmos/block.go [hunk 1]
```diff
@@ -45,73 +54,93 @@ func (s *BlockService) WriteBlock(block *BlockResponse, latest bool) {
...
+func (c *HTTPClient) GetBlock(height *int) (*coretypes.ResultBlock, error) {
+	res := &rpctypes.RPCResponse{}
 
 	hs := ""
 	if height != nil {
 		hs = strconv.Itoa(*height)
 	}
 
-	_, err := c.RPC.R().SetResult(&res).SetError(&resErr).SetQueryParam("height", hs).Get("/block")
+	_, err := c.RPC.R().SetResult(&res).SetError(res).SetQueryParam("height", hs).Get("/block")
```
Comment (kaladinlight) [inline go/pkg/cosmos/block.go:85]:
```suggestion
	_, err := c.RPC.R().SetResult(res).SetError(res).SetQueryParam("height", hs).Get("/block")
```

File: go/coinstacks/thorchain/api/handler.go [hunk 2]
```diff
@@ -66,3 +145,34 @@ func (h *Handler) ParseFee(tx signing.Tx, txid string, denom string) cosmos.Valu
...
+	}
+
+	t := &cosmos.Tx{
+		BaseTx: api.BaseTx{
+			BlockHash:   &block.Hash,
+			BlockHeight: block.Height,
+			Timestamp:   block.Timestamp,
+		},
+		Index:         -1, // synthetic transactions don't have a real tx index
+		Fee:           cosmos.Value{Amount: "2000000", Denom: "rune"},
```
Comment (kaladinlight) [inline go/coinstacks/thorchain/api/handler.go:163]: TODO: need to actually get fees somehow...

Comment (kaladinlight) [reply go/coinstacks/thorchain/api/handler.go]: https://github.com/shapeshift/unchained/pull/512/commits/e4b5c8f287d57c4a56385a41e602f7ea748831cb

File: go/go.sum [hunk 3]
```diff
@@ -25,10 +25,6 @@ cloud.google.com/go v0.84.0/go.mod h1:RazrYuxIK6Kb7YrzzhPoLmCVzl7Sup4NrbKPg8KHSU
 cloud.google.com/go v0.87.0/go.mod h1:TpDYlFy7vuLzZMMZ+B6iRiELaY7z/gJPaqbMx6mlWcY=
```
Comment (asamere) [inline go/go.sum:25]: i have diffs in go.sum after `go mod tidy`

File: go/coinstacks/thorchain/api/handler.go [hunk 4]
```diff
@@ -50,6 +53,97 @@ func (h *Handler) GetAccount(pubkey string) (api.Account, error) {
...
+		for _, e := range events {
+			switch v := e.(type) {
+			case *thorchain.EventFee:
+				if txid == v.TxID {
+					return thorchain.CoinToValue(v.Coins)
+				}
+			}
+		}
+
+		return cosmos.Value{Amount: "2000000", Denom: "rune"}
```
Comment (asamere) [inline go/coinstacks/thorchain/api/handler.go:68]: check out `NativeTransactionFee` from `/thorchain/constants` on the lcd. it's not subject to change often so def not a blocker

**Review by asamere — APPROVED** [review]

**Pull request merged**